### PR TITLE
First stage of the prototype of "REGAPIC" - a gRPC adapter to use REST/JSON

### PR DESCRIPTION
**Pull request by jskeet**
This is not ready to merge, but should give an idea of how I expect it to hang together.

(This PR is so that it's off my local machine, and folks can have a look at the progress.)

### CODE REVIEW

File: Google.Api.Gax.Grpc/Rest/GrpcRestAdapter.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Api.Gax.Grpc.Rest
+{
+    /// <summary>
+    /// Implementation of <see cref="GrpcAdapter"/> that uses HTTP/1.1 and JSON,
+    /// but via a gRPC <see cref="CallInvoker"/>.
+    /// </summary>
+    public sealed class RestGrpcAdapter : GrpcAdapter
```
Comment (chrisdunelm) [inline Google.Api.Gax.Grpc/Rest/GrpcRestAdapter.cs:19]: The file is named `GrpcRestAdapter`, whereas this class is named `RestGrpcAdapter`; I assume that at least one of these should be changed ;)

File: Google.Api.Gax.Grpc/Rest/RestChannel.cs [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+            HttpStatusCode.InternalServerError => StatusCode.Internal,
+            HttpStatusCode.NotFound => StatusCode.NotFound,
+            HttpStatusCode.Conflict => StatusCode.FailedPrecondition,
+            // TODO: Others!
+            _ => StatusCode.Unknown
+        };
+
+        private Metadata GetTrailers(Task<HttpResponseMessage> httpResponseTask)
+        {
+            if (httpResponseTask.IsCompleted)
```
Comment (chrisdunelm) [inline Google.Api.Gax.Grpc/Rest/RestChannel.cs:129]: I think this needs a `!`, if the exception message is to be believed.

**Review by chrisdunelm — COMMENTED** [review]
I don't have all the context here, but the approach looks reasonable.

Just a couple of minor comments.

**Comment by jskeet** [comment]
Doh - thanks for both of those. Addressed.

File: Google.Api.Gax.Grpc/Rest/RestChannelCredentialsConfigurator.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+ */
+
+using Grpc.Core;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Google.Api.Gax.Grpc.Rest
+{
+    class RestChannelCredentialsConfigurator : ChannelCredentialsConfiguratorBase
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/RestChannelCredentialsConfigurator.cs:15]: There's another class named the same and very similar in CredentialExtensions. I believe this one here is not being used, only the one in CredentialExtensions.

File: Google.Api.Gax.Grpc/Rest/CredentialExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+        private sealed class RestChannelCredentialsConfigurator : ChannelCredentialsConfiguratorBase
+        {
+            internal AsyncAuthInterceptor Interceptor { get; private set; }
+
+            // TODO: Validate that we're okay to discard the ChannelCredentials.
+            // This isn't very clearly documented...
+            public override void SetCompositeCredentials(object state, ChannelCredentials channelCredentials, CallCredentials callCredentials)
+            {
+                Interceptor = callCredentials.ToAsyncAuthInterceptor();
```
Comment (amanda-tarafa) [inline Google.Api.Gax.Grpc/Rest/CredentialExtensions.cs:52]:
```suggestion
                var configurator = ChannelCredentialsOnlyConfigurator();
                // If SetComposite is called here, then this fails with NotSupportedException.
                channelCredentials.InternalPopulateConfiguration(configurator, null);
                
                Interceptor = callCredentials.ToAsyncAuthInterceptor();
```

**Review by amanda-tarafa — COMMENTED** [review]
I've left a couple of comments as well, this is neat, and simpler than I thought it would be.

File: Google.Api.Gax.Grpc/Rest/RestChannelCredentialsConfigurator.cs (hunk 3 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/RestChannelCredentialsConfigurator.cs:15]: Deleted.

File: Google.Api.Gax.Grpc/Rest/CredentialExtensions.cs (hunk 4 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Rest/CredentialExtensions.cs:52]: I don't follow this... will discuss when we get closer to actually merging into the main branch.

**Comment by jskeet** [comment]
Merged into regapic-branch. Will all need reviewing again before merging into the main branch, but I've tried to clean up some of the comments.

**Pull request merged**